bigbluebutton-html5/imports/startup/server/customVirtualBackground.js

```diff
@@ -0,0 +1,160 @@
+}
+
+if (existsSync(CUSTOM_BACKGROUND_DIR)) {
+  const files = readdirSync(CUSTOM_BACKGROUND_DIR, 'utf-8');
+
+  if (files.length > 0) {
+    const configs = files.map((file) => {
+      try {
+        // Background
+        const bgPath = `${CUSTOM_BACKGROUND_DIR}/${file}`;
+        copyFileSync(bgPath, `${virtualBgDir}/${file}`);
+        spawnSync('jpegoptim', ['-q', `${virtualBgDir}/${file}`]);
+        const bgData = readFileSync(`${virtualBgDir}/${file}`, 'utf-8');
+        const bgStats = statSync(`${virtualBgDir}/${file}`);
+        const bgHash = createHash('sha1')
+          .update(bgData)
+          .digest('hex');
+        
+        Meteor.settings.public.virtualBackgrounds.fileNames.push(file);
+
+        // Thumbnail
+        const thumbPath = `${CUSTOM_BACKGROUND_DIR}/thumbnails/${file}`;
+        if (existsSync(thumbPath)) {
+          copyFileSync(thumbPath, `${virtualBgDir}/thumbnails/${file}`);
+        } else {
+          spawnSync('convert', [
+            `${CUSTOM_BACKGROUND_DIR}/${file}`,'-resize', '50x50^', '-gravity', 'center',
+            '-extent', '50x50', `${virtualBgDir}/thumbnails/${file}`
+          ]);
+        }
+        spawnSync('jpegoptim', ['-q', `${virtualBgDir}/thumbnails/${file}`]);
+        const thumbData = readFileSync(`${virtualBgDir}/thumbnails/${file}`, 'utf-8');
+        const thumbStats = statSync(`${virtualBgDir}/thumbnails/${file}`);
+        const thumbHash = createHash('sha1')
+          .update(thumbData)
+          .digest('hex');
+
+        // Config
+        return [
+          {
+            custom: true,
+            path: `app/resources/images/virtual-backgrounds/thumbnails/${file}`,
+            where: 'client',
+            type: 'asset',
+            cacheable: false,
+            url: `/resources/images/virtual-backgrounds/thumbnails/${file}`,
+            size: thumbStats.size,
+            hash: thumbHash,
+            sri: null,
+          },
+          {
+            custom: true,
+            path: `app/resources/images/virtual-backgrounds/${file}`,
+            where: 'client',
+            type: 'asset',
+            cacheable: false,
+            url: `/resources/images/virtual-backgrounds/${file}`,
+            size: bgStats.size,
+            hash: bgHash,
+            sri: null,
+          }
+        ];
+      } catch (error) {
+        rmSync(`${virtualBgDir}/${file}`, { force: true });
+        rmSync(`${virtualBgDir}/thumbnails/${file}`, { force: true });
+        // eslint-disable-next-line no-console
+        console.error('Error adding custom virtual background', `${CUSTOM_BACKGROUND_DIR}/${file}`);
+        // eslint-disable-next-line no-console
+        console.error('Files removed.');
+        // eslint-disable-next-line no-console
+        console.error(error);
+        return null;
+      }
+    });
+
+    if (env === 'production') {
+      const programJsonPath = realpathSync(`${meteorRoot}/../programs/web.browser/program.json`);
```

Comment:
Do you really have to modify a file that is shipped by the packages?

Reply:
I don't know how used this file is, but it stores information about public resources. So I don't know if that's really necessary. I did this as a guarantee.